hexedit.1
.TH HEXEDIT 1 "12 July 1998"
.SH NAME
hexedit \- view and edit files in hexadecimal or in ASCII
.SH SYNOPSIS
.I hexedit
[\-s | \-\-sector] [\-m | \-\-maximize] [\-l<n> | \-\-linelength <n>] [\-h | \-\-help] [filename]
.SH DESCRIPTION
.LP
.I hexedit
shows a file both in ASCII and in hexadecimal. The file can be a device
as the file is read a piece at a time. You can modify the file and search through it.
.SH OPTIONS
.TP
.I "\-s, \-\-sector"
Format the display to have entire sectors.
.TP
.I "\-m, \-\-maximize"
Try to maximize the display.
.TP
.I "\-\-color"
Display colors.
This feature is only available if your operating system supports it.
.TP
.I "\-l<n>, \-\-linelength <n>"
Explicitly set the number of bytes to display per line to <n>.
.TP
.I "\-h, \-\-help"
Show the usage.
.SH COMMANDS (quickly)
.SS Moving
.nf
<, > :  go to start/end of the file
Right:  next character
Left:   previous character
Down:   next line
Up:     previous line
Home:   beginning of line
End:    end of line
PUp:    page forward
PDown:  page backward
.fi
.SS Miscellaneous
.nf
F2:     save
F3:     load file
F1:     help
Ctrl-L: redraw
Ctrl-Z: suspend
Ctrl-X: save and exit
Ctrl-C: exit without saving

Tab:    toggle hex/ascii
Return: go to
Backspace: undo previous character
Ctrl-U: undo all
Ctrl-S: search forward
Ctrl-R: search backward
.fi
.SS Cut&Paste
.nf
Ctrl-Space: set mark
Esc-W:  copy
Ctrl-Y: paste
Esc-Y:  paste into a file
Esc-I:  fill
.fi
.SH COMMANDS (full and detailed)
o \fIRight-Arrow\fR, \fILeft-Arrow\fR, \fIDown-Arrow\fR, \fIUp-Arrow\fR \- move the cursor.
.br
o \fICtrl+F\fR, \fICtrl+B\fR, \fICtrl+N\fR, \fICtrl+P\fR \- move the cursor.
.br
o \fICtrl+Right-Arrow\fR, \fICtrl+Left-Arrow\fR, \fICtrl+Down-Arrow\fR, \fICtrl+Up-Arrow\fR \- move n times the cursor.
.br
o \fIEsc+Right-Arrow\fR, \fIEsc+Left-Arrow\fR, \fIEsc+Down-Arrow\fR, \fIEsc+Up-Arrow\fR \- move n times the cursor.
.br
o \fIEsc+F\fR, \fIEsc+B\fR, \fIEsc+N\fR, \fIEsc+P\fR \- move n times the cursor.
.br
o \fIHome\fR, \fICtrl+A\fR \- go the beginning of the line.
.br
o \fIEnd\fR, \fICtrl+E\fR \- go to the end of the line.
.br
o \fIPage up\fR, \fIEsc+V\fR, \fIF5\fR \- go up in the file by one page.
.br
o \fIPage down\fR, \fICtrl+V\fR, \fIF6\fR \- go down in the file by one page.
.br
o \fI<\fR, \fIEsc+<\fR, \fIEsc+Home\fR \- go to the beginning of the file.
.br
o \fI>\fR, \fIEsc+>\fR, \fIEsc+End\fR \- go to the end of the file (for regular files that have a size).
.br
o \fICtrl+Z\fR \- suspend hexedit.
.br
o \fICtrl+U\fR, \fICtrl+_\fR, \fICtrl+/\fR \- undo all (forget the modifications).
.br
o \fICtrl+Q\fR \- read next input character and insert it (this is useful for
inserting control characters and bound keys).
.br
o \fITab\fR, \fICtrl+T\fR \- toggle between ASCII and hexadecimal.
.br
o \fI/\fR, \fICtrl+S\fR \- search forward (in ASCII or in hexadecimal, use \fITAB\fR to change).
.br
o \fICtrl+R\fR \- search backward.
.br
o \fICtrl+G\fR, \fIF4\fR \- go to a position in the file.
.br
o \fIReturn\fR \- go to a sector in the file if \fI\-\-sector\fR is used, otherwise go
to a position in the file.
.br
o \fIEsc+L\fR \- display the page starting at the current cursor position.
.br
o \fIF2\fR, \fICtrl+W\fR \- save the modifications.
.br
o \fIF1\fR, \fIEsc+H\fR \- help (show the man page).
.br
o \fICtrl+O\fR, \fIF3\fR \- open another file
.br
o \fICtrl+L\fR \- redisplay (refresh) the display (usefull when your terminal screws up).
.br
o \fIBackspace\fR, \fICtrl+H\fR \- undo the modifications made on the previous byte.
.br
o \fIEsc+Ctrl+H\fR \- undo the modifications made on the previous bytes.
.br
o \fICtrl+Space\fR, \fIF9\fR \- set mark where cursor is.
.br
o \fIEsc+W\fR, \fIDelete\fR, \fIF7\fR \- copy selected region.
.br
o \fICtrl+Y\fR, \fIInsert\fR, \fIF8\fR \- paste (yank) previously copied region.
.br
o \fIEsc+Y\fR, \fIF11\fR \- save previously copied region to a file.
.br
o \fIEsc+I\fR, \fIF12\fR \- fill the selection with a string
.br
o \fIEsc+T\fR \- truncate the file at the current location
.br
o \fICtrl+C\fR \- unconditional quit (without saving).
.br
o \fIF10\fR, \fICtrl+X\fR \- quit.
.PP
For the \fIEsc\fR commands, it sometimes works to use \fIAlt\fR instead of
\fIEsc\fR. Funny things here (especially for froggies :) egrave = Alt+H ,
ccedilla = Alt+G, Alt+Y = ugrave.
.br
.SS Modeline
At the bottom of the display you have the modeline (copied from emacs). As in
emacs, you have the indications --, ** and %% meaning unmodified, modified and
read-only. Then you have the name of the file you're currently editing. Next to
it is the current position of the cursor in the file followed by the total file
size. The total file size isn't quite correct for devices.
.br
While in --sector mode, it shows the sector the cursor is in.
.SS Editing
You can edit in ASCII or in hexadecimal. You can switch between the two with
\fITab\fR. When the file is read-only, you can't edit it. When trying to edit a
read-only file, a message (```File is read-only''') tells you it is non-writable.
.br
The modifications are shown in bold until they are saved.
The modeline indicates whether you have modified the file or not.
.br
When editing in hexadecimal, only \fI0\fR,\fI1\fR,...,\fI9\fR,
\fIa\fR,\fIb\fR,...,\fIf\fR, \fIA\fR,\fIB\fR,...\fIF\fR are legal.
Other keys are unbound. The first time you hit an unbound key, the help pops up.
It won't pop again unless you call the help directly (with \fIF1\fR).
.br
When editing in ascii, you can find it difficult to enter characters like
\fI/\fR which are bound to a function. The solution is to use the quoted insert
function \fICtrl+Q\fR, the key after the quoted insert function is not processed
by \fIhexedit\fR (like emacs' quoted-insert, or like the \\ character in C).
.SS Searching
You can search for a string in ASCII or in hexadecimal. You can switch
between the two with \fITab\fR. If the string is found, the cursor is moved to
the beginning of the matching location. If the search failed, a message (```not
found''') tells you so. You can cancel the search by pressing a key.
.br
The search in hexadecimal is a bit confusing. You must give a hexadecimal string
with an even number of characters. The search can then be done byte by byte. If
you want to search a long number (eg: a 32 bit number), you must know the
internal representation of that number (little/big endian problem) and give it
the way it is in memory. For example, on an Intel processor (little endian), you
must swap every bytes: 0x12345678 is written 0x78563412 in memory and that's the
string you must give to the search engine.
.br
Before searching you are asked if you want to save the changes, if the file is
edited.
.PP
For more sophisticated search, see Volker Schatz's patch at
<http://www.volkerschatz.com/unix/homebrew.html#hexedit>.
.SS Selecting, copying, pasting, filling
First, select the part of the buffer you want to copy: start setting the mark
where you want. Then go to the end of the area you want to copy (you can use the
go to function and the search functions). Then copy it. You can then paste the
copied area in the current file or in another file.
.PP
You can also fill the selected area with a string or a character: start choosing
the block you want to fill in (set mark then move to the end of the block), and
call the fill function (\fIF12\fR). \fIhexedit\fR ask you the string you want to
fill the block with.
.br
The code is not tuned for huge filling as it keeps the modifications in memory
until you save them. That's why \fIhexedit\fR will warn you if you try to fill
in a big block.
.PP
When the mark is set, the selection is shown in reverse mode.
.br
Be aware that the copied area contains the modifications done at the time of the
copy. But if you undo the modifications, it does not change the content of the
copy buffer. It seems obvious but it's worth saying.
.SS Scrolling
The scrolling is different whether you are in \fI\-\-sector\fR mode or not. In
normal mode, the scrolling is line by line. In sector mode, the scrolling is
sector by sector. In both modes, you can force the display to start at a given
position using \fIEsc+L\fR.
.SH SEE ALSO
od(1), hdump(1), hexdump(1), bpe(1), hexed(1), beav(1).
.SH AUTHOR
Pixel (Pascal Rigaux) <pixel@rigaux.org>, 
.br
Home page is <http://rigaux.org/>.
.SH UNRESTRICTIONS
.I hexedit
is Open Source; anyone may redistribute copies of 
.I hexedit
to
anyone under the terms stated in the GNU General Public License.
.PP
You can find
.I hexedit
at 
.br
<https://github.com/pixel/hexedit/archive/1.3.tar.gz>
.SH TODO
Anything you think could be nice...
.SH LIMITATIONS
There are problems with the curses library given with Redhat 5.0 that make
\fIhexedit\fR think the terminal is huge. The result is that hexedit is
not usable.
.PP
The shortcuts work on some machines, and not on others. That's why there are
many shortcuts for each function. The Ctrl+Arrows and the Alt+. do not work
work as they should most of the time. On SUNs, you must do Ctrl+V-Ctrl+V instead 
of Ctrl+V (!); and the Alt key is the diamond one.
.PP
While searching, it could be interesting to know which position the search has
reached. It's always nice to see something moving to help waiting.
.PP
The hexadecimal search could be able to search modulo 4 bits instead of 8 bits.
Another feature could be to complete padd odd length hexadecimal searches with
zeros.
.SH BUGS
I have an example where the display is completly screwed up. It seems to be a
bug in ncurses (or maybe in xterm and rxvt)?? Don't know if it's me using
ncurses badly or not... It seems to happen when \fIhexedit\fR leaves only one
space at the end of the lines... If anyone has a (or the) solution, please tell
me!
.PP
If you have any problem with the program (even a small one), please do report it
to me. Remarks of any kind are also welcome.
.PP


Changes
	- terminal resizing support

novembre 2016:
	- mark cursor position in both HEX and ASCII
	- skip whitespace when parsing hex-strings

march 2014:
	- fix small but in DIE() which was leaving the terminal in a bad state
	- added --linelength <n> / -l<n> command line option

february 2013:
	- fix displaying sector number when above 2^31
	- fix potential file descriptor leak (thanks to Rich Burridge)
	- add DESTDIR support to the makefiles
	- preprocessor flags should use CPPFLAGS, not CFLAGS
	- fix a small issue in mymemmem/mymemrmem when HAVE_MEMMEM/HAVE_MEMRMEM is not defined

september 2005:
	- version 1.2.12
	- colored (fruit salad) display built by default (if use_default_colors available), 
	  available through option --color (based on a patch from Yuri D'Elia)

september 2005:
	- version 1.2.11
	- add --enable-colors to build and have a colored (fruit salad) display 
	- allow entering goto "position" and "sector" in hexadecimal or not
	- when prompting, propose previously entered value, and give ability to modify it
	- fix setting mark after scrolling
	- some cleanup

may 2004:
	- version 1.2.10
	- handle getnstr() not defined (needed for SGI IRIX 6.5.22)

april 2004:
	- version 1.2.9 (brown paper bag version)
	- fix searching larger than 3 characters long strings

january 2004:
	- version 1.2.8
	- replace the unsafe getstr() with getnstr() (thanks to Richard Chadderton)

august 2003:
	- version 1.2.7
	- check the presence of <sys/mount.h> before including it
	- fix build on Digital UNIX 4.0f (remove -Wall)
	- use the full width of the terminal (the trick is to stop relying on
	printing "\n", but using move() instead)
	- fix displaying the offsets at beginning of each line on big endian
	boxes (thanks to Martin Schaffner)

june 2003:
	- version 1.2.6
	- fix core dump when searching backward (thanks to Jean Delvare)
          (the culprit is memrmem, but also fixing provided memrchr)

june 2003:
	- version 1.2.5
	- fix build on Solaris (which doesn't have BLKGETSIZE and all compilers don't have -W)

march 2003:
	- version 1.2.4
	- use BLKGETSIZE to try to get the size of a device
	- do not die horribly when accessing after the end
	- fix provided memrchr() (thanks to Yue Luo)
	- fix displaying after searching >32bit files (thanks to Paul Mattal)

june 2002:
	- version 1.2.3
	- add some KEY_HOME KEY_END (^[[7~ and ^[[8~)
	- fix for HAVE_MEMMEM
	- fix my memrchr not behaving the same as libc's
	- fix the use of memrchr to behave as libc's
	- use "(void)" in prototype of functions having no parameters
	- call "raw()" when help() is over

october 2001:
	- version 1.2.2
	- yet again some s/int/INT/ (mainly in the "edited" list manipulation)

june 2001:
	- version 1.2.1
	- replace llseek with lseek + use of _FILE_OFFSET_BITS=64
	- memrchr&memmem now exists in some glibc's, so define it only if not provided
	- cleanup basename declaration
	- various cleanup in configure.in, now using autoheader
	- a few bug fixes

april 99: 
	- version 1.1.1
	- replaced lots of int by INT which 64bits
	- replaced lseek by llseek (!! portability=linux :( !!)

11 january 99: (:-<>, it's been a long time)

	- added many shortcuts so that F1..F12 and some others works on more
	terms (xterm, rxvt, sgi's winterm...)
	- removed the scanw that's badly handled on sgi (replaced by getstr +
	sscanf)

24 septembre 98: now displays the reached offset while searching

10 september 98: hexedit 1.0.0 release

08 september 98: Oliver Kiddle's <opk@thoth.u-net.com> changes:
	- changed code where a function that returns a void was returning the
          result of another function. This avoids errors in stricter compilers
	- used an if where ?: was being used to the left of an assignment
	- created Makefile.in and configure.in to replace Makefile
        - added alternative strdup, strerror and memcmp for lacking systems 

26 august 98: hexedit 0.9.5 release

26 august 98: removed the total file size in modeline for devices

19 august 98: Evin Robertson's changes:
	- split hexedit.c into display.c, file.c, hexedit.c, hexedit.h,
	  interact.c, mark.c, misc.c, page.c, search.c
	- Makefile changes - now defaults to installing in /usr/local/bin
	- scrolls one line at a time (but not in sector mode)
	- allow appending to the file
	- errors during writing are now non-fatal
	- show total file size in modeline (not really correct for devices)
	- gives better error messages when functions return -1
	- added newlines to the end of fatal error messages
	- changed sizeof(bufferAttr) to sizeof(*bufferAttr)
	- grammatical corrections to the man page
	- removed dashes so the display is 16 bytes wide on an 80 column
	  screen
	- backspace now moves back an entire byte in hex mode (I thought the
	  previous behavior was inconsistent; it moved the cursor one
	  nybble, but undid an entire byte)
	- refuse to open directories
	- disallow using return to go past the end of the file
	- added truncating ability (M-T)

07 august 98: ability to fill the selection with a string

06 august 98: forgot to mark all the things while doing them. Here is a list of
	what's been added:

	- make the display start at the current cursor position
	- go to a sector when in --sector mode
	- added a modeline
	- ability to open another file. Nice for copy&paste between files
	- you're no more forced to save the file per page. This means the
	cut&paste can now always be undone. It's much simplier that way (idea
	from Evin C Robertson, thanks)
	- the backspace acts as a small undo of the previous character. It's
	nice when typing (another idea from Evin C Robertson)
	- the modified-unmodified bytes are not shown in bold (eg: you replace
	'a' by 'a')

26 july 98: added a TODO file

26 july 98: added save copied region in a file

26 july 98: added search backward

25 july 98: added copy&paste functions

25 july 98: you can now select a part of the buffer

21 july 98: modified bytes are shown in bold.


hexedit.c
/* hexedit -- Hexadecimal Editor for Binary Files
   Copyright (C) 1998 Pixel (Pascal Rigaux)

   This program is free software; you can redistribute it and/or modify
   it under the terms of the GNU General Public License as published by
   the Free Software Foundation; either version 2, or (at your option)
   any later version.

   This program is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.

   You should have received a copy of the GNU General Public License
   along with this program; if not, write to the Free Software
   Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.*/
#include "hexedit.h"


/*******************************************************************************/
/* Global variables */
/*******************************************************************************/
INT lastEditedLoc, biggestLoc, fileSize;
INT mark_min, mark_max, mark_set;
INT base, oldbase;
int normalSpaces, cursor, cursorOffset, hexOrAscii;
int cursor, blocSize, lineLength, colsUsed, page;
int isReadOnly, fd, nbBytes, oldcursor, oldattr, oldcursorOffset;
int sizeCopyBuffer, *bufferAttr;
char *progName, *fileName, *baseName;
unsigned char *buffer, *copyBuffer;
typePage *edited;

char *lastFindFile = NULL, *lastYankToAFile = NULL, *lastAskHexString = NULL, *lastAskAsciiString = NULL, *lastFillWithStringHexa = NULL, *lastFillWithStringAscii = NULL;


const modeParams modes[LAST] = {
  { 8, 16, 256 },
  { 4, 0, 0 },
};
modeType mode = maximized;
int colored = FALSE;

const char * const usage = "usage: %s [-s | --sector] [-m | --maximize] [-l<n> | --linelength <n>]"
#ifdef HAVE_COLORS 
     " [--color]"
#endif 
     " [-h | --help] filename\n";


/*******************************************************************************/
/* main */
/*******************************************************************************/
int main(int argc, char **argv)
{
  progName = basename(argv[0]);
  argv++; argc--;

  for (; argc > 0; argv++, argc--) 
    {
      if (streq(*argv, "-s") || streq(*argv, "--sector"))
	mode = bySector;
      else if (streq(*argv, "-m") || streq(*argv, "--maximize")) {
	mode = maximized;
	lineLength = 0;
      }
#ifdef HAVE_COLORS
      else if (streq(*argv, "--color"))
	colored = TRUE;
#endif
      else if (strbeginswith(*argv, "-l") || strbeginswith(*argv, "--linelength")) {
	if (strbeginswith(*argv, "-l") && strlen(*argv) > 2)
	  lineLength = atoi(*argv + 2);
	else {
	  argv++; argc--;
	  lineLength = atoi(*argv);
	}
      } else if (streq(*argv, "--")) {
	argv++; argc--;
	break;
      } else if (*argv[0] == '-')

	DIE(usage)
      else break;
    }
  if (argc > 1) DIE(usage);

  init();
  if (argc == 1) {
    fileName = strdup(*argv);
    openFile();
  }
  initCurses();
  if (fileName == NULL) {
    if (!findFile()) {
      exitCurses();
      DIE("%s: No such file\n");
    }
    openFile();
  }
  readFile();
  do display();
  while (key_to_function(getch()));
  quit();
  return 0; /* for no warning */
}



/*******************************************************************************/
/* other functions */
/*******************************************************************************/
void init(void)
{
  page = 0; /* page == 0 means initCurses is not done */
  normalSpaces = 3;
  hexOrAscii = TRUE;
  copyBuffer = NULL;
  edited = NULL;
  #if HAVE_MAGIC
    initMagic();
  #endif
}

void quit(void)
{
  #if HAVE_MAGIC
    exitCurses();
  #endif
  freeMagic();
  free(fileName);
  free(buffer);
  free(bufferAttr);
  FREE(copyBuffer);
  discardEdited();
  FREE(lastFindFile); FREE(lastYankToAFile); FREE(lastAskHexString); FREE(lastAskAsciiString); FREE(lastFillWithStringHexa); FREE(lastFillWithStringAscii);
  exit(0);
}





